Performance Improvements - Logging

RohitRanjanMS opened the pull request:
This PR addresses the issue [#2842](https://github.com/Azure/azure-webjobs-sdk/issues/2824). 
Enabling Application Insights has significant performance and throughput overhead. Function supports adding context to log/metric telemetry using BeginScope(). Each call to BeginScope() creates another level of nesting, so that logs/metrics produced within the scope can be tagged with information about where they came from. While this is done extensively in host, customers can create scopes as well (.Net In-Proc) . Nick did an [investigation](https://github.com/Azure/azure-webjobs-sdk/issues/2824) to highlight the performance overhead of using scopes. This is primarily due to multiple Dictionaries that get allocated to support scope. One at the time of passing the scope attributes, another when these attributes are received and wrapped in ReadOnlyDictionary. There’s another set of allocation when the current scope is built where-in it recursively iterate through the nested scopes.

Improvements in this PR
1.	`ApplicationInsightsLogger.cs`
a.	Removed `BeginScope` from the `Log()` method. All the `known properties` are passed as param and added to the telemetry property. This reduces the level of nesting and also reduces the cost of building `scopeInfo`.
b.     Reduced the numer of times `GetMergedStateDictionaryOrNull` is called.
c.	Removed Linq from `ApplyScopeProperties()`. 
d.     Saved some boxing/unboxing and made type casting more efficient.

2.	`LogLevelEnumHelper .cs`
Enum helper to convert `enum` into `string`. This uses a cached string array. No allocation and very efficient compared to `ToString()`. 

3.	`WebJobsTelemetryInitializer.cs`
a.     Restricted the call to `GetMergedStateDictionaryOrNull()` for only `DependencyTelemetry`. 
b.     Category, LogLevel, EventId and EventName are set in `ApplicationInsightsLogger.Log()` method.

4.	`DictionaryLoggerScope.cs`
a.	`Push()` - If state is a dictionary, then use it otherwise build a new dictionary.
b.	`GetMergedStateDictionaryOrNull()` – Return null if `Current` is `null`. 
c.      Cached the `scopeInfo` and return from cache if exist. The cache is tied to the current scope.
e.	`_itemCount` – to avoid dictionary resizing.

**Benchmarking**:
Memory Allocation:
- Dictionary<string, object> allocation is down by ~99.75%. 
- Resize is almost negligible.
 
![image](https://user-images.githubusercontent.com/90008725/224613186-65c4a74f-d4d2-4920-addb-6c82781f36e9.png)

- Req/sec and avg response time is down by ~8%
<img width="972" alt="Screenshot 2023-03-10 112101" src="https://user-images.githubusercontent.com/90008725/224613394-426eb9af-4653-4ef3-a33e-dcef99a370b4.png">

<img width="853" alt="image" src="https://user-images.githubusercontent.com/90008725/214270085-9ea5180a-5a75-4c66-9705-f4009b2f94ab.png">

Execution Time
![image](https://user-images.githubusercontent.com/90008725/224613164-b22e92a9-8dea-4639-b1b7-9c35a89cf5a4.png)

RohitRanjanMS committed:
Initial Commit.

RohitRanjanMS committed:
Fixed Eventid

RohitRanjanMS committed:
More changes

RohitRanjanMS committed:
More changes

RohitRanjanMS committed:
Fixed current scope

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Extensions/LogLevelEnumHelper.cs:
```diff
@@ -0,0 +1,38 @@
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static string ToStringOptimized(LogLevel enumValue)
```
Consider making this an extension method so we can call `logLevel.ToStringOptimized()`.

Also, can we use a switch statement?

``` CSharp
public static string ToStringOptimized(this LogLevel logLevel)
{
    return logLevel switch =>
    {
        LogLevel.Trace => "Trace",
        LogLevel.Debug => "Debug",
        // ... etc
       _ => logLevel.ToString(), // or throw here?
    }
}
```

NickCraver replied:
Agreed - there's no way these change - recommend simplifying to like the core framework has: https://github.com/dotnet/runtime/blob/9b76c28567640e4cbe0d20e18b765b8f1a47473f/src/libraries/Microsoft.Extensions.Logging.Console/src/JsonConsoleFormatter.cs#L92 This is both faster and has zero startup time cost.

RohitRanjanMS replied:
Done

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:
```diff
@@ -33,7 +35,7 @@ internal class ApplicationInsightsLogger : ILogger
         internal const string MetricMaxKey = "max";
         internal const string MetricStandardDeviationKey = "standarddeviation";
-
+        
```
nit: remove extra whitespace

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:
```diff
@@ -165,17 +165,29 @@ private void LogMetric(IEnumerable<KeyValuePair<string, object>> values)
+        private static void ApplyScopeProperties(ITelemetry telemetry)
         {
-            var scopeProperties = DictionaryLoggerScope.GetMergedStateDictionaryOrNull();
+            var scopeProperties = DictionaryLoggerScope.GetMergedStateDictionaryOrNull();            
```
nit: remove extra whitespace
```suggestion
            var scopeProperties = DictionaryLoggerScope.GetMergedStateDictionaryOrNull();
```

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:
```diff
@@ -165,17 +165,29 @@ private void LogMetric(IEnumerable<KeyValuePair<string, object>> values)
+        private static void ApplyScopeProperties(ITelemetry telemetry)
         {
-            var scopeProperties = DictionaryLoggerScope.GetMergedStateDictionaryOrNull();
+            var scopeProperties = DictionaryLoggerScope.GetMergedStateDictionaryOrNull();            
```
I am curious why we do not use `ISupportExternalScope`? Would it be possible for us to switch to it, or is that work for another time?

RohitRanjanMS replied:
I did consider IExternalScopeProvider and we use this in all the other loggers as well. In AI logger, scope is built more than once, and I am able to cache it in the `DictionaryLoggerScope` to save allocation. Additionally, we may later remove AI logger in favor of OTel.

NickCraver replied:
We should consider changing this though - we throw away anything from `GetMergedStateDictionaryOrNull` that's not `!SystemScopeKeys.Contains` or `ScopeKeys.FunctionName` below. For example, a minimal allocation reduction (as well as dictionary size optimization) would be to pass a static predicate to `GetMergedStateDictionaryOrNull` limiting the return and only needing to exclude FunctionName in the foreach below. Happy to brainstorm much more efficient ways to do this!

RohitRanjanMS replied:
I removed the `State` from the `DictionaryLoggerScope `.  I am building the aggregate scope every time a new scope is pushed. `GetMergedStateDictionaryOrNull ` simply returns the current scope.

<img width="1307" alt="image" src="https://user-images.githubusercontent.com/90008725/216011746-623671d0-c87e-47e0-a13c-a7927970ff0d.png">

<img width="1110" alt="image" src="https://user-images.githubusercontent.com/90008725/216013978-02ca5731-ff2c-4649-af89-c16c32f9760f.png">

RohitRanjanMS replied:
I will push the code after testing.

brettsam replied:
Agree with investigating external scopes -- I just commented elsewhere (I didn't see this comment chain) -- but this DictionaryLoggerScope code was written before IExternalScopeProvider stuff existed (it came in .NET Core 2.1; I added this stuff in 2.0 if I recall). What happens if we switch to that and abandon this thing? I know we access it in the initializer, but we may be able to hack something together to test it at least.

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:
```diff
@@ -236,35 +252,54 @@ private void LogTrace(LogLevel logLevel, IEnumerable<KeyValuePair<string, object
+            {
+                if (eventId.Id != 0)
+                {
+                    properties[LogConstants.EventIdKey] = Convert.ToString(eventId.Id);
```
Can you explain why `Convert.ToString(int)` over `int.ToString()`? Or rather `int.ToString(CultureInfo.InvariantCulture)`, which can be made into an extension metho `ToStringInviariant()`.

RohitRanjanMS replied:
There's not much gain here, just my preferred way to convert to string. I will add InvariantCulture.

<img width="1013" alt="image" src="https://user-images.githubusercontent.com/90008725/214463317-2456c941-f3b6-47dc-a906-7bb7cf58f8eb.png">

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:
```diff
@@ -236,35 +252,54 @@ private void LogTrace(LogLevel logLevel, IEnumerable<KeyValuePair<string, object
+        {
+            properties[LogConstants.CategoryNameKey] = _categoryName;
+            properties[LogConstants.LogLevelKey] = LogLevelEnumHelper.ToStringOptimized(logLevel);
+            if (eventId != null)
```
If you are interested, pattern matching can be used to reduce the nesting here

```csharp
if (eventId { Id: != 0 })
{
}

if (eventId { Name: not null })
{
}
```

RohitRanjanMS replied:
eventId is a struct, null check for eventId is redundant.

NickCraver replied:
I'm confused in reviewing the resolved conversations here - are there changes not pushed up perhaps?

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:
```diff
@@ -236,35 +252,54 @@ private void LogTrace(LogLevel logLevel, IEnumerable<KeyValuePair<string, object
+            {
+                stringValue = value;
+            }
+            if (objectValue is DateTime date)
```
Note sure if you intentionally didn't use `else if` here.
```suggestion
            else if (objectValue is DateTime date)
```

RohitRanjanMS replied:
I missed it.

NickCraver replied:
FWIW a switch statement would be much more concise here with the `else if` change in play (I can't see this reviewing latest on GitHub though)

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:
```diff
@@ -307,40 +342,46 @@ private void LogFunctionResultAggregate(IEnumerable<KeyValuePair<string, object>
                         break;
-                    default:
-                        if (value.Value is TimeSpan)
+                    default:                        
```
nit: whitespace
```suggestion
                    default:
```

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/DictionaryLoggerScope.cs:
```diff
@@ -17,7 +18,10 @@ private DictionaryLoggerScope(IReadOnlyDictionary<string, object> state, Diction
         }
-
+        // Cache merged dictionary. Invalidate cache on push/dispose 
+        private IDictionary<string, object> _currentScope;
```
nit: place fields before the ctor.

RohitRanjanMS committed:
Update src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs

Co-authored-by: Jacob Viau <javia@microsoft.com>

NickCraver commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:
```diff
@@ -130,7 +123,14 @@ private void LogMetric(IEnumerable<KeyValuePair<string, object>> values)
+                        else
+                        {
+                            telemetry.Name = entry.Value.ToString();
+                        }
```
Can unconditionally call .ToString() here to be a bit cleaner - it'll just get inlined as a direct ref return anyhow: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/String.cs#L541

RohitRanjanMS replied:
Done.

NickCraver commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:
```diff
@@ -308,39 +343,45 @@ private void LogFunctionResultAggregate(IEnumerable<KeyValuePair<string, object>
             foreach (KeyValuePair<string, double> metric in metrics)
             {
-                _telemetryClient.TrackMetric($"{functionName} {metric.Key}", metric.Value);
+                _telemetryClient.TrackMetric($"{functionName} {metric.Key}", metric.Value, properties);
```
Is `metric.Key` a finite set of values? Allocations per-concat here would be something we can eliminate if so

RohitRanjanMS replied:
yeah, but functionName varies as well.

NickCraver commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Initializers/WebJobsTelemetryInitializer.cs:
```diff
@@ -49,47 +49,61 @@ public void Initialize(ITelemetry telemetry)
+                        LogLevel? logLevel = scopeProps.GetValueOrDefault<LogLevel?>(LogConstants.LogLevelKey);
+                        if (logLevel != null)
+                        {
+                            telemetryContext.Properties[LogConstants.LogLevelKey] = logLevel.Value.ToString();
```
Can use the switched .ToString method here to avoid an allocation

RohitRanjanMS replied:
Done.

NickCraver commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Initializers/WebJobsTelemetryInitializer.cs:
```diff
@@ -49,47 +49,61 @@ public void Initialize(ITelemetry telemetry)
+                        if (category != null)
+                        {
+                            telemetryContext.Properties[LogConstants.CategoryNameKey] = category;
+                        }
```
FWIW can simplify these with your new format, e.g. 
```cs
if (scopeProps.GetValueOrDefault<string>(LogConstants.CategoryNameKey) is string category)
```

RohitRanjanMS replied:
Done

NickCraver commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Initializers/WebJobsTelemetryInitializer.cs:
```diff
@@ -49,47 +49,61 @@ public void Initialize(ITelemetry telemetry)
+                        int? eventId = scopeProps.GetValueOrDefault<int?>(LogConstants.EventIdKey);
+                        if (eventId != null && eventId.HasValue && eventId.Value != 0)
+                        {
+                            telemetryContext.Properties[LogConstants.EventIdKey] = eventId.Value.ToString();
```
Should this be culture aware as well?

RohitRanjanMS replied:
Done.

NickCraver reviewed (commented):
Changes are looking good - My main concern is around the `GetMergedStateDictionaryOrNull` shifts - have we done memory profiling on that tree in particular? (I see the total profiling - nice! - I'm only asking about the remainder) I worry with the new approach we may be adding dictionary allocations overall. I think we can scope this in some places but that's messy too - IMO we need another way to go about doing this all together, but I'm not sure where else it's still used looking only at the diff scope of this PR.

Overall: I think there's probably a round of changes here I'm not seeing in GitHub's current version given resolved comments, happy to take a peek at latest if helpful - nice work!

RohitRanjanMS committed:
Fixed if else

RohitRanjanMS committed:
Merge branch 'roranjan/AILoggerPerfImprovement' of https://github.com/Azure/azure-webjobs-sdk into roranjan/AILoggerPerfImprovement

RohitRanjanMS committed:
Removed state from DictionaryScope

RohitRanjanMS committed:
Pattern matching

RohitRanjanMS committed:
Removed space

RohitRanjanMS committed:
Removed unused namespace

brettsam commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/DictionaryLoggerScope.cs:
```diff
@@ -12,16 +12,16 @@ internal class DictionaryLoggerScope
     {
         private static AsyncLocal<DictionaryLoggerScope> _value = new AsyncLocal<DictionaryLoggerScope>();
 
-        private DictionaryLoggerScope(IReadOnlyDictionary<string, object> state, DictionaryLoggerScope parent)
```
Have you looked into getting rid of this type entirely? I wrote this years ago before ExternalScopes were supported and have changed every other instance of logging to use that rather than this. The only holdout is how we access the scope in the TelemetryInitializer, but I never did enough research to figure out how to get around that. 

I'd be curious how much improvement we get by using the built-in ScopeProvider, which I assume is optimized for performance...

I'd be interested to see performance reports if this was used (even if it is fairly hacked together for now).

RohitRanjanMS replied:
I did try `IExternalScopeProvider` but it didn't work.
1. As you mentioned, we need the scope in the `TelemetryInitializer` for Dependencies. All telemetry items flow through the `ApplicationInsightsLogger` except dependencies. `TelemetryInitializer` is the only way I can think of to enrich dependencies with additional information.
2. IExternalScopeProvider calls an internal `BeginScope` instead of the [one](https://github.com/Azure/azure-webjobs-sdk/blob/dev/src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs#:~:text=public%20IDisposable-,BeginScope,-%3CTState%3E(TState%20state)) provided by us. We need this for `StartTelemetryIfFunctionInvocation`. `ApplicationInsightsLogger` is the only logger that uses `BeginScope`.

I did some benchmarking against IExternalScopeProvider and it's very efficient and works like magic. Unfortunately, couldn't use it here. 
The proposed changes are efficient as well, we are saving almost 99.74% allocation even with all the complexities.

jviau replied:
@RohitRanjanMS can you share your code with the `IExternalScopeProvider` implementation?

RohitRanjanMS replied:
sure, I don't have a commit, but I can make a quick change and push to a branch.

RohitRanjanMS replied:
I created a branch that uses [`IExternalScopeProvider`](https://github.com/Azure/azure-webjobs-sdk/compare/roranjan/IExternalScopeProvider?expand=1). I don't see any issue with `BeginScope()` now and the only thing that's stopping us from using this is [`WebJobsTelemetryInitializer`](https://github.com/Azure/azure-webjobs-sdk/blob/roranjan/IExternalScopeProvider/src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Initializers/WebJobsTelemetryInitializer.cs#:~:text=IDictionary%3Cstring%2C%20object%3E%20scopeProps%20%3D%20new%20Dictionary%3Cstring%2C%20object%3E()%3B//DictionaryLoggerScope.GetMergedStateDictionaryOrNull()%3B).

RohitRanjanMS replied:
I did some [benchmarking ](https://github.com/RohitRanjanMS/LoggerFactoryScopeProviderBenchmarking/tree/master/LoggerFactoryScopeProviderBenchmarking) using `LoggerFactoryScopeProvider`. While the performance is almost same, memory allocation is significantly less.

<img width="614" alt="image" src="https://user-images.githubusercontent.com/90008725/226382842-adda4e03-8e15-4422-9a40-31942e1338cb.png">

jviau replied:
Thank you for the branch. I understand your issue now of needing access to the log scope - or rather to a telemetry property bag for multiple forms of telemetry (traces and logs here). I believe this is a concept solved by OTel Baggage, but we don't have access to it here. I don't really like the idea of rolling our own scope and having this discrepancy in how this _one_ `ILogger` works - but addressing this is not that straight forward.

As for the allocation difference you see - yes, your sample branch is allocating an extra dictionary. You can remove that perf hit by avoiding the dictionary allocation and attaching scope data directly to where it needs to go within the `ForEachScope` delegate. Additionally, you want to use the `state` parameter to pass in any variables you access within the `ForEachScope` delegate to avoid function closures. With those 2 changes you should see significantly less allocations.

RohitRanjanMS committed:
Fixed null handling.

NickCraver reviewed (commented):
Current code looks good from an efficiency standpoint - I'm only reviewing in text so limited here and profiles are very much appreciated!

My only suggestion here would be: do we want to beef up the testing slightly specifically on scopes/expected values? There's enough of a change here I'm not sure existing tests cover all the cases that can happen with overrides, what if the value set in an override is null, etc - just from a 1:1 behavior with the old for nested properties.

RohitRanjanMS commented:
> Current code looks good from an efficiency standpoint - I'm only reviewing in text so limited here and profiles are very much appreciated!
> 
> My only suggestion here would be: do we want to beef up the testing slightly specifically on scopes/expected values? There's enough of a change here I'm not sure existing tests cover all the cases that can happen with overrides, what if the value set in an override is null, etc - just from a 1:1 behavior with the old for nested properties.

I will have a look at test cases and add more if we don't have enough coverage.

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/ApplicationInsightsLogger.cs:
```diff
@@ -165,17 +158,28 @@ private void LogMetric(IEnumerable<KeyValuePair<string, object>> values)
+        private static void ApplyScopeProperties(ITelemetry telemetry)
         {
-            var scopeProperties = DictionaryLoggerScope.GetMergedStateDictionaryOrNull();
+            var scopeProperties = DictionaryLoggerScope.GetMergedStateDictionaryOrNull();            
```
nit: remove extra whitespace

RohitRanjanMS replied:
Fixed.

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/DictionaryLoggerScope.cs:
```diff
@@ -12,16 +12,16 @@ internal class DictionaryLoggerScope
+        {
+            CurrentScope = currentScope;
+            Parent = parent;
+        }
```
nit: add new line after `}`

RohitRanjanMS replied:
Done.

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Extensions/LogLevelExtension.cs:
```diff
@@ -0,0 +1,38 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static string ToStringOptimized(this LogLevel logLevel)
+        {
+            switch (logLevel)
```
style nit: consider switch expression.

``` CSharp
return logLevel switch
{
    LogLevel.Trace => "Trace",
    ...
    _ => logLevel.ToString(CultureInfo.InvariantCulture),
};
```

RohitRanjanMS replied:
This feature is not available in C# 7.3

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/Initializers/WebJobsTelemetryInitializer.cs:
```diff
@@ -49,53 +50,61 @@ public void Initialize(ITelemetry telemetry)
-                if (category != null)
+                // this could be telemetry tracked in scope of function call - then we should apply the logger scope
+                // or RequestTelemetry tracked by the WebJobs SDK or AppInsight SDK - then we should apply Activity.Tags
+                if (scopeProps != null && scopeProps.Count > 0)
```
small suggestion: 
```suggestion
                if (scopeProps?.Count > 0)
```

RohitRanjanMS replied:
Done.

jviau commented on src/Microsoft.Azure.WebJobs.Logging.ApplicationInsights/DictionaryLoggerScope.cs:
```diff
@@ -37,54 +37,62 @@ public static DictionaryLoggerScope Current
+                IDictionary<string, object> stateValues;
+                // Convert this to a dictionary as we have scenarios where we cannot have duplicates.
+                // In this case, if there are dupes, the later entry wins.
                 stateValues = new Dictionary<string, object>();
```
If you want even more performance out of this scope implementation, I recommend looking at [LoggerFactoryScopeProvider](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Logging/src/LoggerFactoryScopeProvider.cs) and basically following the `IExternalScopeProvider` pattern. It does a great job at avoiding extra allocations. For example here you are allocating an extra dictionary. And then `BuildCurrentScope` is going to allocate _another_ dictionary yet. `IExternalScopeProvider` avoids these allocations by not collapsing/processing the scope upfront, but rather letting consumers iterate through the scope values and they must perform type checking and consuming each scope level appropriately - and they also need to decide how to handle duplicates.

Not a blocker for this PR, but I thought I would mention this as the primary focus of this is performance.

RohitRanjanMS replied:
The dictionary allocation in the `Push` method is only if the state is `IEnumerable`. If it's already a dictionary, then there is no allocation. In all our code path we use a dictionary for state. `IEnumerable` is just to make sure we don't break if any customer is using it for scope.

jviau reviewed (approved):
Small nits and some perf suggestions.

My primary concern is the way this makes our one `ApplicationInsightsLogger` unique. I worry that we have two ways log scopes are done, and it is not clear if you need to use `ILogger.BeginScope` or this other `DictionaryLoggerScope`. I get that we are trying to bring log scope properties over to distributed tracing (AI dependencies here). Did we explore if `Activity.Tags` would work here? Or is there some other prescribed pattern for having a general "telemetry scope" instead of a log scope?

I am approving this as the PR goal is performance and the `DictionaryLoggerScope` deviation (which is my primary concern) predates this PR.

RohitRanjanMS committed:
Optimizing for the first `BeginScope` invocation.

RohitRanjanMS committed:
Fixed code

RohitRanjanMS committed:
Added tests to validate overriding in DictionaryLoggerScope.

RohitRanjanMS committed:
Fixed int/double casting.

RohitRanjanMS committed:
Add scope properties for logs emitted using TelemetryClient.
e.g.,
TelemetryClient client = new TelemetryClient();
            client.InstrumentationKey = "<>";
            client.TrackTrace("msg");

RohitRanjanMS committed:
Fixed spacing
